axelor-production/src/main/java/com/axelor/apps/production/service/batch/BatchFillSopRealValues.java

```diff
@@ -0,0 +1,180 @@
+    ProductCategory productCategory = sop.getProductCategory();
+    for (SopLine sopLine : sop.getSopLineList()) {
+      try {
+        computeRealSalesAndSalesGap(company, productCategory, sopLine);
+        computeRealProduction(company, productCategory, sopLine);
+        computeRealStock(company, productCategory, sopLine);
+        incrementDone();
+      } catch (Exception e) {
+        incrementAnomaly();
+        TraceBackService.trace(e, ExceptionOriginRepository.SOP, batch.getId());
+      }
+    }
+  }
+
+  protected void computeRealSalesAndSalesGap(
+      Company company, ProductCategory productCategory, SopLine sopLine) {
+    Period period = sopLine.getPeriod();
+    sopLine.setSopRealSales(
+        saleOrderLineRepo
+            .all()
+            .filter(
+                "self.saleOrder.statusSelect IN (3,4) AND self.product.productCategory = :productCategory AND self.saleOrder.company = :company AND self.saleOrder.confirmationDateTime BETWEEN :startDate AND :endDate")
+            .bind("productCategory", productCategory)
+            .bind("company", company)
+            .bind("startDate", period.getFromDate())
+            .bind("endDate", period.getToDate())
+            .fetchStream()
+            .map(SaleOrderLine::getInTaxTotal)
+            .reduce(BigDecimal.ZERO, BigDecimal::add));
+
+    // compute % sales gap
+    computeSopSalesGap(sopLine);
+  }
+
+  protected void computeRealProduction(
+      Company company, ProductCategory productCategory, SopLine sopLine) {
+    Period period = sopLine.getPeriod();
+    sopLine.setSopRealManuf(
+        manufOrderRepo
+            .all()
+            .filter(
+                "self.statusSelect = :statusSelect AND self.product.productCategory = :productCategory AND self.company = :company AND self.realEndDateT BETWEEN :startDate AND :endDate")
+            .bind("statusSelect", ManufOrderRepository.STATUS_FINISHED)
+            .bind("company", company)
+            .bind("productCategory", productCategory)
+            .bind("startDate", period.getFromDate())
+            .bind("endDate", period.getToDate())
+            .fetchStream()
+            .map(ManufOrder::getQty)
+            .reduce(BigDecimal.ZERO, BigDecimal::add));
+  }
+
+  protected void computeRealStock(
+      Company company, ProductCategory productCategory, SopLine sopLine) {
+    Period period = sopLine.getPeriod();
+    sopLine.setSopRealStock(
+        stockLocationLineHistoryRepo
+            .all()
+            .filter(
+                "self.stockLocationLine.stockLocation.company = :company AND self.stockLocationLine.product.productCategory = :productCategory AND self.dateT < :endDate")
+            .bind("company", company)
+            .bind("productCategory", productCategory)
+            .bind("endDate", period.getToDate())
+            .fetchStream()
+            .map(history -> history.getQty().multiply(history.getWap()))
+            .reduce(BigDecimal.ZERO, BigDecimal::add));
```

Comment:
Same as above, use directly a query (createQuery method) to compute the sum instead of using a fetchStream